netty/src/main/java/io/grpc/netty/ConfigurableX509TrustManager.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.netty;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import io.grpc.netty.TlsOptions.VerificationAuthType;
+import java.net.Socket;
+import java.security.KeyStore;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLParameters;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.X509ExtendedTrustManager;
+
+/**
+ * ConfigurableX509TrustManager is a highly configurable class that allows users choose different
+ * level of peer checking mechanisms, as well as some customized check. It could also be used to
+ * reload trust certificate bundle client/server uses.
+ */
+public class ConfigurableX509TrustManager extends X509ExtendedTrustManager {
+
+  private TlsOptions tlsOptions;
+
+  public ConfigurableX509TrustManager(TlsOptions tlsOptions) {
+    this.tlsOptions = checkNotNull(tlsOptions, "tlsOptions");
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] x509Certificates, String s, Socket socket)
+      throws CertificateException {
+
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] x509Certificates, String s, SSLEngine sslEngine)
+      throws CertificateException {
+    checkTrusted(x509Certificates, s, sslEngine, false);
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] x509Certificates, String s)
+      throws CertificateException {
+
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] x509Certificates, String s, Socket socket)
+      throws CertificateException {
+
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] x509Certificates, String s, SSLEngine sslEngine)
+      throws CertificateException {
+    checkTrusted(x509Certificates, s, sslEngine, true);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] x509Certificates, String s)
+      throws CertificateException {
+
+  }
+
+  @Override
+  public X509Certificate[] getAcceptedIssuers() {
+    return new X509Certificate[0];
```

Comment:
Implementing it will cause one test to fail, and I think the cause is probably https://github.com/netty/netty/blob/netty-4.1.48.Final/handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslServerContext.java#L153-L164. It looks like if setting this. Netty internally will use it to do some additional checks, which is not the behavior I want. I might want `checkClientTrusted` and `checkServerTrusted` to cover all the trust behaviors we might have during the authentication.

Reply:
Hmm... That's sort of worrisome. Go ahead and put a comment here saying why it is empty.

Reply:
Actually... this looks like it is only for certificate chain verification. What does it break?

Reply:
I've summarized my findings in this Gist: https://gist.github.com/ZhenLian/50c73f146493ac2fe9746d196082679b.
If I changed `getAcceptedIssuers` to the one named `change.java` in my Gist, one of the tests I wrote would fail, with the error message named `failure.log`.
That is actually an expected behavior under my current implementation. The failing test is basically to test under the scenario that server chooses to `SkipAllVerification`, and client sends a bad certificate, in mTLS. This should be OK because server chooses to skip all verification(in real situations, we want application to provide additional custom checks, but since this is test, we just test the simple behavior). But this test won't pass if we have additional check in `getAcceptedIssuers`.